client/structs/csi.go

```diff
@@ -0,0 +1,121 @@
+		return true
+	default:
+		return false
+	}
+}
+
+// CSIVolumeAccessMode indicates how a volume should be used in a storage topology
+// e.g whether the provider should make the volume available concurrently.
+type CSIVolumeAccessMode string
+
+const (
+	CSIVolumeAccessModeUnknown CSIVolumeAccessMode = ""
+
+	CSIVolumeAccessModeSingleNodeReader CSIVolumeAccessMode = "single-node-reader-only"
+	CSIVolumeAccessModeSingleNodeWriter CSIVolumeAccessMode = "single-node-writer"
+
+	CSIVolumeAccessModeMultiNodeReader       CSIVolumeAccessMode = "multi-node-reader-only"
+	CSIVolumeAccessModeMultiNodeSingleWriter CSIVolumeAccessMode = "multi-node-single-writer"
+	CSIVolumeAccessModeMultiNodeMultiWriter  CSIVolumeAccessMode = "multi-node-multi-writer"
+)
+
+// ValidCSIVolumeAccessMode checks to see that the provided access mode is a valid,
+// non-empty access mode.
+func ValidCSIVolumeAccessMode(accessMode CSIVolumeAccessMode) bool {
+	switch accessMode {
+	case CSIVolumeAccessModeSingleNodeReader, CSIVolumeAccessModeSingleNodeWriter,
+		CSIVolumeAccessModeMultiNodeReader, CSIVolumeAccessModeMultiNodeSingleWriter,
+		CSIVolumeAccessModeMultiNodeMultiWriter:
+		return true
+	default:
+		return false
+	}
+}
+
+// CSIVolumeMountOptions contains the mount options that should be provided when
+// attaching and mounting a volume with the CSIVolumeAttachmentModeFilesystem
+// attachment mode.
+type CSIVolumeMountOptions struct {
+	// Filesystem is the desired filesystem type that should be used by the volume
+	// (e.g ext4, aufs, zfs). This field is optional.
+	Filesystem string
+
+	// MountFlags contain the mount options that should be used for the volume.
+	// These may contain _sensitive_ data and should not be leaked to logs or
+	// returned in debugging data.
+	// The total size of this field must be under 4KiB.
+	MountFlags []string
+}
+
+type ClientCSIControllerAttachVolumeRequest struct {
+	PluginName string
+
+	// The ID of the volume to be used on a node.
+	// This field is REQUIRED.
+	VolumeID string
+
+	// The ID of the node. This field is REQUIRED. This must match the NodeID that
+	// is fingerprinted by the target node for this plugin name.
+	NodeID string
+
+	// AttachmentMode indicates how the volume should be attached and mounted into
+	// a task.
+	AttachmentMode CSIVolumeAttachmentMode
+
+	// AccessMode indicates the desired concurrent access model for the volume
+	AccessMode CSIVolumeAccessMode
+
+	// MountOptions is an optional field that contains additional configuration
+	// when providing an AttachmentMode of CSIVolumeAttachmentModeFilesystem
+	MountOptions *CSIVolumeMountOptions
+
+	// ReadOnly indicates that the volume will be used in a readonly fashion. This
+	// only works when the Controller has the PublishReadonly capability.
+	ReadOnly bool
+}
+
+func (c *ClientCSIControllerAttachVolumeRequest) ToCSIRequest() *csi.ControllerPublishVolumeRequest {
+	if c == nil {
+		return &csi.ControllerPublishVolumeRequest{}
+	}
+
```

Comment:
We deliberately don't thread through topology information to the CSI Client yet. This will come when I get an initial loop of volume attach/detach together because some of the wiring will be a little tricky, and warrants it's own PR.